analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/EmptyNullableValueAccess.cs

```diff
@@ -850,6 +850,69 @@ async Task WithAsyncAwait()
     }
 }
 
+class Boxing
+{
+    void EmptyNullable()
+    {
+        int? nullable = null;
+        object implicitBoxed = nullable;
+        _ = (int)implicitBoxed;             // Compliant, true null-dereference -> no nullable value access
+        var explicitBoxed = (object)nullable;
+        _ = (int)explicitBoxed;             // Compliant
+        _ = (int)(object)(int?)(null);      // Compliant
+        _ = (int)(object)(null as int?);    // Compliant
+    }
+
+    void NullableRoundTrip()
+    {
+        int? nullable = null;
+        object implicitBoxed = nullable;
+        int? unboxedNullable = (int?)implicitBoxed;
+        _ = unboxedNullable.Value;          // Noncompliant, empty
+    }
+
+    void NonEmptyNullable()
+    {
+        int? nullable = 42;
+        object implicitBoxed = nullable;
+        _ = (int)implicitBoxed;             // Compliant
+        var explicitBoxed = (object)nullable;
+        _ = (int)explicitBoxed;             // Compliant
+    }
+
+    void ImplicitAndExplicitConversion()
+    {
+        int? intNullable = 42;
+        object boxedInt = intNullable;
+        _ = (short)intNullable;             // Compliant, cast exception from int unboxing -> no nullable value access
+    }
+
+    void ForeachImplictConversion()
+    {
+        foreach (object boxed in new int?[] { null })
+        {
+            _ = (int)boxed;                 // Compliant
+        }
+    }
+
+    void CollectionImplicitBoxing()
+    {
+        foreach (var boxed in new object[] { null as int? })
```

Comment:
We already have this example at another place right:
```cs
        foreach (int boxed in new int?[] { null })
```

Reply:
I think we should have test cases for for each for the following cases:

### explicit conversion from nullable to non-nullable

* explicit conversion from int? to int (e.g. `foreach(int i in new int?[] { 1, null })`)
* lifted explicit conversions (e.g. `foreach(int i in new long?[] { 1, null })`). Here there are actually two explicit conversions at stake: one form nullable to non-nullable and one from wider type (long) to narrow type (int).

These are the cases I asked for originally.

### boxing conversions from nullable to object

`foreach (object o in new int?[] { null })`

These are [implicit conversions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/conversions#1028-implicit-reference-conversions). This kind of operation should always succeed, so I don't see the need for much test coverage here.

### unboxing into nullable

`foreach (int? i in new object[] { null, 1, "Boom" })`

This has nothing to do with the rule itself. No need to cover it with tests.

### stackalloc

I don't see how this makes a difference. It is just another kind of location to enumerate. You can also use other kinds of pattern-based foreach collections, like ImmutableArrays Enumerator. It doesn't make a difference for the test cases.

In general, I would look at the C# spec rather than the IL generated. The IL generated needs to follow the spec behavior-wise. I don't see any violations from that in the examples you have given.

The sections needed are
* [Conversion from a nullable value type to an underlying type](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types#conversion-from-a-nullable-value-type-to-an-underlying-type)
* [Lifted operators](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types#lifted-operators)
* [foreach](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/statements/iteration-statements#the-foreach-statement)

Reply:
Ok, good, it's clear now, thanks for this analysis! 

So:

### explicit conversion from nullable to non-nullable
Added a new method `ForeachExplicitConversion` with the two examples

### boxing conversions from nullable to object
The conversion indeed succeeds, but the cast in the body of the foreach does not.
`(int)boxed` fails with `NullReferenceException` at the first item of the collection which is `null`.
I find non-obvious at first glance that:
- if boxing happens, you end up with a `NullReferenceException`, raised by the runtime in the boxing operation, and S3655 does not kick in;
- whereas, if boxing doesn't happen, you end up with an `InvalidOperationException`, raised by the framework in `Nullable<T>.Value`, and S3655 does kick in.
I wanted to document this difference in behavior, by keeping these two test scenarios `ForeachImplictConversion` and `ForeachExplicitConversion`.

### unboxing into nullable
Agreed, I haven't added it.

### stackalloc
Indeed, it just change the memory layout. 
It just seemed interesting, since stack allocation in general can affect boxing. 
Removed.